### PR TITLE
Vaiduoklis ubuntu.lt service

### PR DESCRIPTION
**Pull request by mgedmin**
I have not attempted to run this with ansible-playbook yet.

### CODE REVIEW

**Comment by mgedmin** [comment]
I have tried this now:

```
vaiduoklis.pov.lt          : ok=28   changed=8    unreachable=0    failed=0    skipped=3    rescued=0    ignored=0   
```

but the service itself failed:

```
● ubuntult.service - ubuntu.lt gunicorn daemon
     Loaded: loaded (/etc/systemd/system/ubuntult.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Sat 2021-05-08 13:53:25 EEST; 59s ago
TriggeredBy: ● ubuntult.socket
    Process: 296394 ExecStart=/opt/ubuntu.lt/venv/bin/gunicorn -w 4 project.wsgi.prod (code=exited, status=1/FAILURE)
   Main PID: 296394 (code=exited, status=1/FAILURE)
     Status: "Gunicorn arbiter booted"
      Tasks: 0 (limit: 9450)
     Memory: 624.0K
     CGroup: /system.slice/ubuntult.service

May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296394]:     time.sleep(0.1)
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296394]:   File "/opt/ubuntu.lt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 242, in handle_chld
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296394]:     self.reap_workers()
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296394]:   File "/opt/ubuntu.lt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 525, in reap_workers
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296394]:     raise HaltServer(reason, self.WORKER_BOOT_ERROR)
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296394]: gunicorn.errors.HaltServer: <HaltServer 'Worker failed to boot.' 3>
May 08 13:53:25 vaiduoklis.pov.lt systemd[1]: ubuntult.service: Main process exited, code=exited, status=1/FAILURE
May 08 13:53:25 vaiduoklis.pov.lt systemd[1]: ubuntult.service: Killing process 296437 (gunicorn) with signal SIGKILL.
May 08 13:53:25 vaiduoklis.pov.lt systemd[1]: ubuntult.service: Killing process 296437 (gunicorn) with signal SIGKILL.
May 08 13:53:25 vaiduoklis.pov.lt systemd[1]: ubuntult.service: Failed with result 'exit-code'.
```

journalctl -u ubuntult sako

```
-- Logs begin at Tue 2021-04-27 11:55:09 EEST, end at Sat 2021-05-08 13:55:46 EEST. --
May 08 13:53:25 vaiduoklis.pov.lt systemd[1]: Starting ubuntu.lt gunicorn daemon...
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296394]: [2021-05-08 13:53:25 +0300] [296394] [INFO] Starting gunicorn 20.1.0
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296394]: [2021-05-08 13:53:25 +0300] [296394] [INFO] Listening at: unix:/run/ubuntult/socket,http://127.0.0.1:8120 (296394)
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296394]: [2021-05-08 13:53:25 +0300] [296394] [INFO] Using worker: sync
May 08 13:53:25 vaiduoklis.pov.lt systemd[1]: Started ubuntu.lt gunicorn daemon.
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]: [2021-05-08 13:53:25 +0300] [296431] [INFO] Booting worker with pid: 296431
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]: [2021-05-08 13:53:25 +0300] [296431] [ERROR] Exception in worker process
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]: Traceback (most recent call last):
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "/opt/ubuntu.lt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:     worker.init_process()
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "/opt/ubuntu.lt/venv/lib/python3.8/site-packages/gunicorn/workers/base.py", line 134, in init_process
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:     self.load_wsgi()
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "/opt/ubuntu.lt/venv/lib/python3.8/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:     self.wsgi = self.app.wsgi()
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "/opt/ubuntu.lt/venv/lib/python3.8/site-packages/gunicorn/app/base.py", line 67, in wsgi
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:     self.callable = self.load()
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "/opt/ubuntu.lt/venv/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:     return self.load_wsgiapp()
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "/opt/ubuntu.lt/venv/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:     return util.import_app(self.app_uri)
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "/opt/ubuntu.lt/venv/lib/python3.8/site-packages/gunicorn/util.py", line 359, in import_app
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:     mod = importlib.import_module(module)
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:     return _bootstrap._gcd_import(name[level:], package, level)
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "<frozen importlib._bootstrap>", line 961, in _find_and_load_unlocked
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]:   File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]: ModuleNotFoundError: No module named 'project.wsgi'
May 08 13:53:25 vaiduoklis.pov.lt gunicorn[296431]: [2021-05-08 13:53:25 +0300] [296431] [INFO] Worker exiting (pid: 296431)
...
```

I wonder if /opt/ubuntult/.env needs to specify PYTHONPATH?  Or maybe the working directory is wrong?

**Comment by mgedmin** [comment]
With the correct WorkingDirectory it runs.  Sort of.

```
mgedmin@vaiduoklis:🐌 /opt/ubuntu.lt/spirit [git:master=] $ curl http://localhost:8120/

<!doctype html>
<html lang="en">
<head>
  <title>Bad Request (400)</title>
</head>
<body>
  <h1>Bad Request (400)</h1><p></p>
</body>
</html>
```

journalctl -u ubuntult says this is because

```
May 08 14:05:54 vaiduoklis.pov.lt gunicorn[298009]: ERROR 2021-05-08 11:05:54,947 exception 298009 140474235561792 Invalid HTTP_HOST header: 'localhost:8120'. You may need to add 'localhost' to ALLOWED_HOSTS.
```

File: websites/ubuntu.lt/templates/ubuntult.service [hunk 1]
```diff
@@ -8,9 +8,9 @@ Type=notify
 User={{ slug }}
 Group=www-data
 RuntimeDirectory={{ slug }}
-WorkingDirectory={{ home }}
+WorkingDirectory={{ path }}
 EnvironmentFile={{ home }}/.env
-ExecStart={{ home }}/venv/bin/gunicorn -w 4 -b 127.0.0.1:8120 project.wsgi.prod
+ExecStart={{ home }}/venv/bin/gunicorn -w 4 project.wsgi.prod
```
Comment (sirex) [inline websites/ubuntu.lt/templates/ubuntult.service:13]: `127.0.0.1:8120` is expected in https://github.com/akllt/infrastructure/blob/master/websites/ubuntu.lt/templates/apache.conf#L53-L54

**Review by sirex — APPROVED** [review]
Currently Apache expect an HTTP server on 8120 port, see my inline comment. Everything looks good.

File: websites/ubuntu.lt/templates/ubuntult.service (hunk 1 above)
Comment (mgedmin) [reply websites/ubuntu.lt/templates/ubuntult.service:13]: For systemd deployments the listening socket address is specified in the `.socket` file, not on the command line here.

Comment (mgedmin) [reply websites/ubuntu.lt/templates/ubuntult.service:13]: If you're curious how that works, see https://www.freedesktop.org/software/systemd/man/systemd.socket.html.  Basically

- you tell systemd what ports to open
- when something connects to that port, systemd starts the associated .service unit
- when starting a service unit, systemd sets a couple of environment variables (LISTEN_FDS and LISTEN_PID)
- gunicorn [reads those environment variables](https://github.com/benoitc/gunicorn/blob/master/gunicorn/systemd.py#L12-L46) and knows which file descriptors, inherited from the parent process (i.e. systemd), contain the listening socket(s).

Comment (sirex) [reply websites/ubuntu.lt/templates/ubuntult.service:13]: Thanks, today I learned a new thing.

**Pull request merged**